# 1. Install and import require packages for EDA

In [193]:
# install require packages
!pip install pandas
!pip install country_converter
!pip install plotly

You should consider upgrading via the 'C:\Users\ad\PycharmProjects\RS\venv\Scripts\python.exe -m pip install --upgrade pip' command.


You should consider upgrading via the 'C:\Users\ad\PycharmProjects\RS\venv\Scripts\python.exe -m pip install --upgrade pip' command.


  Using cached tenacity-8.2.2-py3-none-any.whl (24 kB)


You should consider upgrading via the 'C:\Users\ad\PycharmProjects\RS\venv\Scripts\python.exe -m pip install --upgrade pip' command.


In [194]:
# import require packages

import pandas as pd
import numpy as np
import country_converter as coco
import plotly.express as px


# 2. Load dataset as dataframes

In [132]:
# data paths
book_hist_path = '../book_crossing/book_history.dat'
book_rate_path = '../book_crossing/book_ratings.dat'
items_info_path = '../book_crossing/items_info.dat'
users_info_path = '../book_crossing/users_info.dat'

In [133]:
book_hist_df = pd.read_csv(book_hist_path, delimiter='\t', header=None, low_memory=False, na_values=np.nan)
print(book_hist_df.head())

      0     1         2
0  user  item  accessed
1     1   152         1
2     1   153         1
3     1  2176         1
4     1   154         1


In [134]:
book_rate_df = pd.read_csv(book_rate_path, delimiter='\t', header=None, low_memory=False)
print(book_rate_df.head())

      0     1         2
0  user  item    rating
1     1  6264  7.000000
2     1  4350  7.000000
3     1  6252  5.000000
4     1   202  9.000000


In [168]:
user_info_df = pd.read_csv(users_info_path, delimiter='\t', index_col=None, na_values=np.nan)

In [169]:
print(user_info_df.columns)

Index(['User-ID', 'Location', 'Age'], dtype='object')


In [170]:
user_info_df.head()

User-ID                     Location  Age
0        1  minneapolis, minnesota, usa   24
1        2   san diego, california, usa   20
2        3      novinger, missouri, usa   16
3        4      sonoma, california, usa   34
4        5    berkeley, california, usa   23

In [171]:
items_info_df = pd.read_csv(items_info_path, delimiter='\t')

print(items_info_df.head())

   Book_ID        ISBN                                         Book-Title  \
0        1  0060973129                               Decision in Normandy   
1        2  0393045218                             The Mummies of Urumchi   
2        3  0425176428  What If?: The World's Foremost Military Histor...   
3        4  0452264464               Beloved (Plume Contemporary Fiction)   
4        5  0609804618  Our Dumb Century: The Onion Presents 100 Years...   

       Book-Author Year-Of-Publication                  Publisher  \
0     Carlo D'Este                1991            HarperPerennial   
1  E. J. W. Barber                1999  W. W. Norton &amp Company   
2    Robert Cowley                2000   Berkley Publishing Group   
3    Toni Morrison                1994                      Plume   
4        The Onion                1999         Three Rivers Press   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/0060973129.0...   
1  http://

In [172]:
print(items_info_df.columns)

Index(['Book_ID', 'ISBN', 'Book-Title', 'Book-Author', 'Year-Of-Publication',
       'Publisher', 'Image-URL-S', 'Image-URL-M', 'Image-URL-L'],
      dtype='object')


# 3. EDA

## 3.1  Exploring USER INFO

In [173]:
# get number of users
print(user_info_df.shape)
print(user_info_df.columns)
print(user_info_df.head())

(2946, 3)
Index(['User-ID', 'Location', 'Age'], dtype='object')
   User-ID                     Location  Age
0        1  minneapolis, minnesota, usa   24
1        2   san diego, california, usa   20
2        3      novinger, missouri, usa   16
3        4      sonoma, california, usa   34
4        5    berkeley, california, usa   23


##### >> Total: 2946 users, each user has 3 attributes: 'User-ID', 'Location', 'Age'.

#### 3.1.1. Explore Age from users info

In [174]:
# Print out age values
print(sorted(user_info_df.Age.unique()))

[1, 2, 4, 8, 9, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 75, 76, 77, 79, 80, 81, 82, 83, 90, 100, 101, 103, 116, 136, 148, 168, 201, 204, 239]


##### >> 'Age' columns no have any null values, and have some value very high, more than 100.

##### // As we see, people who are less than 5 years old or more than 90 years old usually can't rate books much, so in this case we will replace them with nan values.

In [175]:
user_info_df.loc[(user_info_df.Age > 90) | (user_info_df.Age < 5), 'Age'] = np.nan

In [176]:
print(user_info_df[(user_info_df.Age > 90) | (user_info_df.Age < 5)])

Empty DataFrame
Columns: [User-ID, Location, Age]
Index: []


In [177]:
print(user_info_df[user_info_df['Age'] == np.nan])

Empty DataFrame
Columns: [User-ID, Location, Age]
Index: []


##### Replace nan values in Age column with mean()

In [178]:
user_info_df['Age'] = user_info_df['Age'].fillna(user_info_df['Age'].mean())

In [179]:
user_info_df.head()

User-ID                     Location   Age
0        1  minneapolis, minnesota, usa  24.0
1        2   san diego, california, usa  20.0
2        3      novinger, missouri, usa  16.0
3        4      sonoma, california, usa  34.0
4        5    berkeley, california, usa  23.0

In [180]:
# convert type 'Age' to integer
user_info_df['Age'] = user_info_df['Age'].astype(int)

In [181]:
user_info_df.head()

User-ID                     Location  Age
0        1  minneapolis, minnesota, usa   24
1        2   san diego, california, usa   20
2        3      novinger, missouri, usa   16
3        4      sonoma, california, usa   34
4        5    berkeley, california, usa   23

#### 3.1.2. Explore User's Location

In [182]:
# explore countries from location
countries = []
user_info_df['Country'] = user_info_df['Location'].apply(lambda x: x.split(',')[-1])
# countries.append(add[-1] for add in addresses)

In [183]:
print(user_info_df.head())

   User-ID                     Location  Age Country
0        1  minneapolis, minnesota, usa   24     usa
1        2   san diego, california, usa   20     usa
2        3      novinger, missouri, usa   16     usa
3        4      sonoma, california, usa   34     usa
4        5    berkeley, california, usa   23     usa


In [184]:
# print out all location from all users
locations = sorted(user_info_df.Location.unique())
countries = sorted(user_info_df.Country.unique())

print(f'We have total: {len(locations)} locations')
print(f'We have total: {len(countries)} countries')

We have total: 1799 locations
We have total: 48 countries


##### >> So, we have 1799 unique locations from 2946 users in dataset

In [185]:
print(user_info_df['Country'].value_counts())

Country
 usa                    2128
 canada                  284
 united kingdom          113
                          72
 australia                54
 germany                  49
 portugal                 35
 spain                    33
 france                   31
 malaysia                 24
 netherlands              22
 austria                  10
 finland                   9
 switzerland               8
 romania                   6
 new zealand               6
 singapore                 5
 philippines               5
 iran                      5
 brazil                    4
 japan                     4
 italy                     3
 ireland                   3
 denmark                   3
 qatar                     2
 china                     2
 belgium                   2
 dominican republic        2
 norway                    2
 poland                    2
 united state              1
 czech republic            1
 sweden                    1
 phillipines               1
 cypru

#### As we see, top 3 countries has most book reviewers: usa (2128), canada (284), united kingdom(113).

In [186]:
print(user_info_df[user_info_df['Country']==''])

      User-ID                    Location  Age Country
5           6          king of prussia, ,   36        
6           7                 berkeley, ,   22        
40         41        lakeland, tennessee,   60        
83         84              murphysboro, ,   37        
178       179  manchester, new hampshire,   51        
...       ...                         ...  ...     ...
2739     2740        ray, michigan usa, ,   45        
2761     2762                  molalla, ,   45        
2871     2872                     reno, ,   29        
2893     2894                     arco, ,   51        
2895     2896            coconut grove, ,   71        

[72 rows x 4 columns]


##### And has 72 people who we miss about their country name

In [201]:
converted_country=coco.convert(names=user_info_df["Country"], to="ISO3")
user_info_df["Country"]=converted_country

In [202]:
country=user_info_df["Country"].value_counts()
user_info_df.loc[user_info_df['Country']=='not found'] = 'USA'
print(country)
print(type((country)))
fig=px.choropleth(locations=country.index,
                  color=country.values,
                  color_continuous_scale=px.colors.sequential.dense,
                  template='plotly_dark',
                  title='Distribution of users by countries')

fig.update_layout(font=dict(size=17, family="Franklin Gothic"))
fig.show()

Country
USA    2205
CAN     284
GBR     113
AUS      54
DEU      49
PRT      35
ESP      33
FRA      31
MYS      24
NLD      22
AUT      10
FIN       9
CHE       8
ROU       6
NZL       6
PHL       5
IRN       5
SGP       5
BRA       4
JPN       4
IRL       3
ITA       3
DNK       3
NOR       2
POL       2
QAT       2
BEL       2
CHN       2
DOM       2
SWE       1
ISL       1
CZE       1
EGY       1
MMR       1
CYP       1
RUS       1
ZAF       1
KWT       1
HKG       1
TWN       1
BGR       1
TTO       1
Name: count, dtype: int64
<class 'pandas.core.series.Series'>


##  3. Explore ITEM INFO

In [203]:
print(items_info_df.head())

   Book_ID        ISBN                                         Book-Title  \
0        1  0060973129                               Decision in Normandy   
1        2  0393045218                             The Mummies of Urumchi   
2        3  0425176428  What If?: The World's Foremost Military Histor...   
3        4  0452264464               Beloved (Plume Contemporary Fiction)   
4        5  0609804618  Our Dumb Century: The Onion Presents 100 Years...   

       Book-Author Year-Of-Publication                  Publisher  \
0     Carlo D'Este                1991            HarperPerennial   
1  E. J. W. Barber                1999  W. W. Norton &amp Company   
2    Robert Cowley                2000   Berkley Publishing Group   
3    Toni Morrison                1994                      Plume   
4        The Onion                1999         Three Rivers Press   

                                         Image-URL-S  \
0  http://images.amazon.com/images/P/0060973129.0...   
1  http://